### DISCUSSION TITLE
load initial namespaces earlier

### DISCUSSION THREAD

**Discussion by Xiphe** [sergiodxa/remix-i18next#72] [Ideas]
As of now, when using this package with a http backend in the client it causes a loading waterfall on the initial load and delays react initialisation a few additional milliseconds.

<img width="813" alt="Screenshot 2022-06-01 at 11 25 24" src="https://user-images.githubusercontent.com/911218/171373753-ea17fc76-4969-468b-a96f-f7eab8220744.png">

I created a [prove of concept](https://github.com/Xiphe/remix-i18n-testyard/tree/t9n-compiler) with a custom i18n-backend in the client and wrapped translation files that eliminates this issue. But I'm not sure how feasible it is. Would anybody be interested in working on this?

**Comment by sergiodxa** [comment]
You could use the detected language to preload the translations rendering a `<link>` tag in your root route.

```tsx
export default function Root() {
  let { locale } = useLoaderData<LoaderData>();
  let { i18n } = useTranslation();
  useChangeLanguage(locale);
  let matches = useMatches()
  let namespaces = [...new Set(matches.flatMap(match => match.handle?.i18n))]

  return (
    <html lang={locale} dir={i18n.dir()}>
      <head>
        <Meta />
        <Links />
        {namespaces.map(namespace => {
          return <link key={namespace} rel="preload" as="fetch" href={`/locales/${locale}/${namespace}.json`} />
        })}
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}

**Reply by sergiodxa** [reply]
There's an undocumented component to do that for your https://github.com/sergiodxa/remix-i18next/blob/8deb188b60455898434bc99381f29d0d8d717f4a/src/react.tsx#L19-L48
But I had issues working with it so I didn't mentioned in the docs.

**Reply by Xiphe** [reply]
Wow, yeah that is a whole lot easier then my current solution! thanks 👍
